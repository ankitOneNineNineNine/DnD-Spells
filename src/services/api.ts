/**
 * Apis List
 * @export api
 */
export const api = {
  spells: {
    fetch: "http://www.dnd5eapi.co/api/spells",
    fetchById: "http://www.dnd5eapi.co/api/spells/{id}",
  },
};
