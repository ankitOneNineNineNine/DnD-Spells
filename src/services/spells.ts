import axios, { AxiosError } from "axios";
import { useQuery } from "react-query";
import { api } from "./api";
import { toastFail } from "./toast";

export interface ISpell {
  index: string;
  name: string;
  url: string;
}
interface ISchool {
  index: string;
  name: string;
  url: string;
}

type IClass = ISpell;
type ISubClass = ISpell;
type IDamageType = ISpell;
export const Componentsenum: { [key: string]: any } = {
  V: "verbal",
  S: "somatic",
  M: "material"
};
export interface ISpellInfo {
  casting_time?: string;
  classes?: IClass[];
  components?: string[];
  concentration?: boolean;
  desc?: string[];
  duration?: string;
  damage?: {
    damage_at_character_level?: { [key: number]: string };
    damage_type?: IDamageType;
  };
  area_of_effect?: {
    size?: number;
    type?: string;
  };
  attack_type?: string;
  higher_level?: string[];
  index?: string;
  level?: 2;
  material?: string;
  name?: string;
  range?: string;
  ritual?: boolean;
  school?: ISchool;
  subclasses?: ISubClass[];
  url?: string;
  _id?: string;
}

export interface IFetchSpellsResponse {
  count: number;
  results: ISpell[];
}

/**
 * Axios Get Call to fetch repos based on search query
 *
 * @returns AxiosResponse<IFetchSpellsResponse>
 */
const fetchSpells = () => () => {
  return axios.get<IFetchSpellsResponse>(api.spells.fetch);
};

/**
 * Axios Get Call to get Spell by Index
 * @param {string} id
 * @returns AxiosResponse<ISpell>
 */
const getSpellByIndex = (id: string) => () => {
  return axios.get<ISpellInfo>(api.spells.fetchById.replace("{id}", id));
};

/**
 * Use Query to Fetch All Spells
 * @export useGetAllSpells
 * @returns UseQueryResult
 */
export const useGetAllSpells = () => {
  return useQuery(api.spells.fetch, fetchSpells(), {
    select: (response) => response.data,
    onError: (
      error: AxiosError<{
        message: string;
      }>
    ) => {
      toastFail(error?.response?.data?.message || "Something Went Wrong");
    }
  });
};

/**
 * Use Query to Fetch Spell By Index
 * @param {string} id
 * @export useGetSpellByIndex
 * @returns UseQueryResult
 */
export const useGetSpellByIndex = (id: string) => {
  return useQuery([api.spells.fetch, id], getSpellByIndex(id), {
    enabled: !!id,
    select: ({ data }) => data,
    onError: (
      error: AxiosError<{
        message: string;
      }>
    ) => {
      toastFail(error?.response?.data?.message || "Something Went Wrong");
    }
  });
};
