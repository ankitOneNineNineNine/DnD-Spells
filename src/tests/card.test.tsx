import React from "react";
import { render, screen } from "@testing-library/react";
import { Card } from "../components/Card/card.components";
import { ISpellInfo } from "../services/spells";

/**
 * Mock Spell Details
 */
let mockSpell: ISpellInfo = {
  concentration: false,
  casting_time: "1 action",
  damage: {
    damage_at_character_level: {},
    damage_type: {
      index: "acid",
      name: "Acid",
      url: "/api/damage-types/acid",
    },
  },
  classes: [{ index: "wizard", name: "Wizard", url: "/api/classes/wizard" }],
  components: ["V", "S", "M"],
  desc: [
    "A shimmering green arrow streaks toward a target w…damage and no damage at the end of its next turn.",
  ],
  duration: "Instantaneous",
  higher_level: [
    "When you cast this spell using a spell slot of 3rd…) increases by 1d4 for each slot level above 2nd.",
  ],
  index: "acid-arrow",
  level: 2,
  material: "Powdered rhubarb leaf and an adder's stomach.",
  name: "Acid Arrow",
  range: "90 feet",
  ritual: false,
  subclasses: [],
  school: {
    index: "evocation",
    name: "Evocation",
    url: "/api/magic-schools/evocation",
  },
  url: "/api/spells/acid-arrow",
  _id: "62fd2c57b749a01481874e0e",
};

describe("card", () => {
  it("should match a snapshot", () => {
    const { container } = render(<Card spell={mockSpell} />);
    expect(container).toMatchSnapshot();
  });

  it("should display title", () => {
    render(<Card spell={mockSpell} />);
    expect(screen.getByText(mockSpell.name!)).toBeTruthy();
  });

  it("should display Empty Title", () => {
    render(<Card />);
    expect(screen.getByRole("heading", { level: 1 })).toBeEmptyDOMElement();
  });
});
