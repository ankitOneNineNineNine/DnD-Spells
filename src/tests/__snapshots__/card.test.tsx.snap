// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`card should match a snapshot 1`] = `
<div>
  <div
    class="cardContainer"
  >
    <h1>
      Acid Arrow
    </h1>
    <h3>
      Powdered rhubarb leaf and an adder's stomach.
    </h3>
    <p>
      <span>
        Casting Time: 
      </span>
      1 action
    </p>
    <p>
      <span>
        Components: 
      </span>
      verbal, somatic, material
    </p>
    <p>
      <span>
        Attack Type: 
      </span>
      Not Mentioned
    </p>
    <p>
      <span>
        Classes That Can Learn Spell : 
      </span>
      Wizard
    </p>
    <p>
      <span>
        Level: 
      </span>
       
      2
    </p>
    <p>
      <span>
         Range of Spell: 
      </span>
      90 feet
    </p>
    <p>
      <span>
        Magic School: 
      </span>
      Evocation
    </p>
    <p>
      <span>
        Higher Level Spell: 
      </span>
      When you cast this spell using a spell slot of 3rd…) increases by 1d4 for each slot level above 2nd.
    </p>
    <h3>
      Description
    </h3>
    <ul
      style="list-style-position: inside;"
    >
      <li>
        A shimmering green arrow streaks toward a target w…damage and no damage at the end of its next turn.
      </li>
    </ul>
  </div>
</div>
`;
